### PR TITLE
Revamp config loading and support datadog.yaml

### PR DESCRIPTION
**Pull request by LotharSee**
Built on top of rebase of #357 this PRs supersedes it.

Exhaustive changes:

From #357:

- support documentation from datadog.yaml
  - support reading from this yaml file and its main attributes, introducing Yaml config structs
  - add apm fields to it under `apm_config`
  - support multiple config sources, merging them together

Added on top:

- clean glide files
- clean the code, refactor the different configuration sources into different merge options
- remove dead options: `APIPayloadBufferMaxSize`
- comment which config options are not documented (they are likely to get removed in future versions)
- remove `agent/trace-agent.ini` and clean `config/README.md`. Now we should only refer to the A5 and A6 example configs.
- deprecate  `-ddconfig` making `-config` the only supported way of configuring the Agent. `-ddconfig` is kept for a bit before getting fully removed.
- simplify Proxy configuration, support the one provided from A6.
- improve coverage and clean (using external files) our config tests

### CODE REVIEW

File: agent/main.go [hunk 1]
```diff
@@ -112,31 +113,37 @@ func runAgent(exit chan struct{}) {
...
+	// trace-agent configuration
+	var agentConf *config.AgentConfig
+	// Agent 6 datadog.yaml config
+	var yamlConf *config.YamlAgentConfig
+	// Agent 5 datadog.conf config
+	var conf *config.File
+	// deprecated Agent 5 trace-agent.ini config
+	var legacyConf *config.File
+
+	if filepath.Ext(opts.configFile) == ".conf" {
```
Comment (truthbk) [inline agent/main.go]: To make the precedence clear, I would probably log a line regarding what config got loaded.

File: config/merge_ini.go [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+	if v, e := conf.GetInt(section, "queue_max_payloads"); e == nil {
+		c.MaxQueuedPayloads = v
+	}
+
+	c.ExponentialBackoff = readExponentialBackoffConfig(conf, section)
+
+	return c
+}
+
+// TODO: maybe this is too many options exposed?
```
Comment (AlexJF) [inline config/merge_ini.go]: Right. Kept them around until we did some experimentation with the new retry scenario. Aaditya's experiments seem to suggest it's working nicely so we could remove them. Although as long as we keep them internal, they don't really affect usability and will allow us to provide an easy hotfix if we ever detect a problem here.

File: agent/main_nix.go [hunk 3]
```diff
@@ -4,20 +4,22 @@ package main
...
 )
 
 func init() {
 	// command-line arguments
-	flag.StringVar(&opts.ddConfigFile, "ddconfig", "/etc/dd-agent/datadog.conf", "Classic agent config file location")
-	// FIXME: merge all APM configuration into dd-agent/datadog.conf and deprecate the below flag
-	flag.StringVar(&opts.configFile, "config", "/etc/datadog/trace-agent.ini", "Trace agent ini config file.")
+	// TODO: load from the .yaml automatically if there
+	flag.StringVar(&opts.configFile, "config", "/etc/dd-agent/datadog.conf", "Datadog Agent config file location")
+	flag.StringVar(&opts.legacyConfigFile, "ddconfig", "/etc/dd-agent/trace-agent.ini", "Deprecated extra configuration option.")
```
Comment (AlexJF) [inline agent/main_nix.go]: This requires changes here:
* https://github.com/DataDog/datadog-trace-agent/blob/9bd73a013f78525cc640be04a36045373f6d5921/Rakefile#L148
* Devenv 
* Devops

File: config/merge_ini.go [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+		c.MaxConnections = v
+	}
+	// undocumented
+	// TODO: remove, should stay internal?
+	if v, e := conf.GetInt("trace.watchdog", "check_delay_seconds"); e == nil {
+		c.WatchdogInterval = time.Duration(v) * time.Second
+	}
+
+	// [trace.writer.*] sections
+	// undocumented, should they stay all internal?
```
[the remark this answers is not in the document]
Comment (LotharSee) [reply config/merge_ini.go]: yup, keeping them here for now for our testing, will remove it in later versions.

File: config/merge_ini.go (hunk 2 above)
Comment (LotharSee) [reply config/merge_ini.go]: yup, keeping them here for now for our testing, will remove it in later versions.

File: agent/main.go (hunk 1 above)
Comment (LotharSee) [reply agent/main.go]: Good one, updated with more logging.

**Pull request merged**